app/scripts/lib/Stores/BaseStore.ts

```diff
@@ -0,0 +1,167 @@
+ * communication to the user about what happened and their best recourse for
+ * handling the situation if the extension cannot gracefully recover.
+ *
+ * 2. Set the current state to the underlying storage system. This should be
+ * implemented in such a way that the current metadata is stored in a separate
+ * key that is tracked by the storage system. This metadata should *not* be a
+ * input to the set method. If the underlying storage system allows for partial
+ * state objects it should be sufficient to pass the data key, which is the
+ * full MetaMask state tree. If not, then the metadata should be supplied by
+ * the storage system itself.
+ *
+ * 3. Provide a method for generating a first time state tree. This method is
+ * implemented as a part of this Abstract class and should not be overwritten
+ * unless future work requires specific implementations for different storage
+ * systems. This method should return a state tree that is the default state
+ * tree for a new install.
+ */
+export abstract class BaseStore {
+  /**
+   * isSupported is a boolean that is set to true if the underlying storage
+   * system is supported by the current browser and implementation.
+   */
+  abstract isSupported: boolean;
+
+  /**
+   * dataPersistenceFailing is a boolean that is set to true if the storage
+   * system attempts to write state and the write operation fails. This is only
+   * used as a way of deduplicating error reports sent to sentry as it is
+   * likely that multiple writes will fail concurrently.
+   */
+  abstract dataPersistenceFailing: boolean;
+
+  /**
+   * mostRecentRetrievedState is a property that holds the most recent state
+   * successfully retrieved from memory. Due to the nature of async read
+   * operations it is beneficial to have a near real-time snapshot of the state
+   * for sending data to sentry as well as other developer tooling.
+   */
+  abstract mostRecentRetrievedState: MetaMaskStorageStructure | null;
+
+  /**
+   * metadata is a property that holds the current metadata object. This object
+   * includes a single key which is 'version' and contains the current version
+   * number of the state tree. This is only incremented via the migrator and in
+   * a well functioning (typical) install should match the latest migration's
+   * version number.
+   */
+  #metadata?: { version: number };
+
+  /**
+   * migrator is a property that holds the migrator instance that is used to
+   * migrate state from one shape to another. This migrator is used to create
+   * the first time state tree.
+   */
+  abstract migrator: Migrator;
+
+  /**
+   * Sets the current metadata. The set method that is implemented in storage
+   * classes only requires an object that is set on the 'data' key. The
+   * metadata key of this class is set on the 'meta' key of the underlying
+   * storage implementation (e.g. chrome.storage.local).
+   */
+  set metadata(metadata: { version: number }) {
```

Comment:
We should consider how to ensure:
* This is always set prior to persistence
* This is correctly loaded from persisted state (or backup persisted state) prior to migrations running
* This cannot be set to the incorrect value (e.g. guard where/how it's set)